worker/provisioner/provisioner_task.go

```diff
@@ -506,27 +461,88 @@ func (task *provisionerTask) findUnknownInstances(stopping []instances.Instance)
 	return unknown, nil
 }
 
-// instancesForDeadMachines returns a list of instances.Instance that represent
-// the list of dead machines running in the provider. Missing machines are
-// omitted from the list.
-func (task *provisionerTask) instancesForDeadMachines(deadMachines []apiprovisioner.MachineProvisioner) []instances.Instance {
-	var instances []instances.Instance
-	for _, machine := range deadMachines {
+func (task *provisionerTask) removeMachines(dead []apiprovisioner.MachineProvisioner) error {
+	// Stop all machines that are dead.
+	stopping := task.instancesForDeadMachines(dead)
+
+	// Find running instances that have no machines associated.
+	unknown, err := task.findUnknownInstances(stopping)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if !task.harvestMode.HarvestUnknown() {
+		task.logger.Infof(
+			"%s is set to %s; unknown instances not stopped %v",
+			config.ProvisionerHarvestModeKey,
+			task.harvestMode.String(),
+			instanceIds(unknown),
+		)
+		unknown = nil
+	}
+
+	if task.harvestMode.HarvestNone() || !task.harvestMode.HarvestDestroyed() {
+		task.logger.Infof(
+			`%s is set to "%s"; will not harvest %s`,
+			config.ProvisionerHarvestModeKey,
+			task.harvestMode.String(),
+			instanceIds(stopping),
+		)
+		stopping = nil
+	}
+
+	if len(stopping) > 0 {
+		task.logger.Infof("stopping known instances %v", stopping)
+	}
+	if len(unknown) > 0 {
+		task.logger.Infof("stopping unknown instances %v", instanceIds(unknown))
+	}
+
+	// It is important that we stop unknown instances before starting
+	// pending ones, because if we start an instance and then fail to
+	// set its InstanceId on the machine.
+	// We don't want to start a new instance for the same machine ID.
+	if err := task.stopInstances(append(stopping, unknown...)); err != nil {
+		return errors.Trace(err)
+	}
+
+	// Remove any dead machines from state.
+	for _, machine := range dead {
+		task.logger.Infof("removing dead machine %q", machine.Id())
+		if err := machine.MarkForRemoval(); err != nil {
+			task.logger.Errorf("failed to remove dead machine %q", machine.Id())
+		}
+		task.removeMachineFromAZMap(machine)
+		task.machinesMutex.Lock()
+		delete(task.machines, machine.Id())
+		task.machinesMutex.Unlock()
```

Comment:
I know this is a mechanical change, but I'd encapsulate this in another method to improve readability.